contrib/envoyproxy/envoy/fakehttp.go

```diff
@@ -0,0 +1,189 @@
+	headers, pseudoHeaders := partitionPeusdoHeaders(res.ResponseHeaders.GetHeaders().GetHeaders())
+
+	if err := checkPseudoResponseHeaders(pseudoHeaders); err != nil {
+		return err
+	}
+
+	status, err := strconv.Atoi(pseudoHeaders[":status"])
+	if err != nil {
+		return fmt.Errorf("error parsing status code %q: %w", pseudoHeaders[":status"], err)
+	}
+
+	for k, v := range headers {
+		w.Header().Set(k, strings.Join(v, ","))
+	}
+
+	w.WriteHeader(status)
+	return nil
+}
+
+// NewRequestFromExtProc creates a new http.Request from an ext_proc RequestHeaders message
+func NewRequestFromExtProc(ctx context.Context, req *extproc.ProcessingRequest_RequestHeaders) (*http.Request, error) {
+	headers, pseudoHeaders := partitionPeusdoHeaders(req.RequestHeaders.GetHeaders().GetHeaders())
+	if err := checkPseudoRequestHeaders(pseudoHeaders); err != nil {
+		return nil, err
+	}
+
+	parsedURL, err := url.Parse(fmt.Sprintf("%s://%s%s", pseudoHeaders[":scheme"], pseudoHeaders[":authority"], pseudoHeaders[":path"]))
+	if err != nil {
+		return nil, fmt.Errorf(
+			"error building envoy URI from scheme %q, from host %q and from path %q: %w",
+			pseudoHeaders[":scheme"],
+			pseudoHeaders[":host"],
+			pseudoHeaders[":path"],
+			err)
+	}
+
+	var remoteAddr string
+	md, ok := metadata.FromIncomingContext(ctx)
+	if ok {
+		remoteAddr = getRemoteAddr(md)
+	}
+
+	var tlsState *tls.ConnectionState
+	if pseudoHeaders[":scheme"] == "https" {
+		tlsState = &tls.ConnectionState{}
+	}
+
+	headers["Host"] = append(headers["Host"], pseudoHeaders[":authority"])
+
+	return (&http.Request{
+		Method:     pseudoHeaders[":method"],
+		Host:       pseudoHeaders[":authority"],
+		RequestURI: pseudoHeaders[":path"],
+		URL:        parsedURL,
+		Header:     headers,
+		RemoteAddr: remoteAddr,
+		TLS:        tlsState,
+	}).WithContext(ctx), nil
+}
+
+type FakeResponseWriter struct {
+	mu      sync.Mutex
+	status  int
+	body    []byte
+	headers http.Header
+}
+
+// Reset resets the FakeResponseWriter to its initial state
+func (w *FakeResponseWriter) Reset() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.status = 0
+	w.body = nil
+	w.headers = make(http.Header)
+}
+
+// Status is not in the [http.ResponseWriter] interface, but it is cast into it by the tracing code
+func (w *FakeResponseWriter) Status() int {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.status
+}
+
+func (w *FakeResponseWriter) WriteHeader(status int) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.status = status
+}
+
+func (w *FakeResponseWriter) Header() http.Header {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.headers
+}
+
+func (w *FakeResponseWriter) Write(b []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.body = append(w.body, b...)
+	return len(b), nil
+}
+
+var _ http.ResponseWriter = &FakeResponseWriter{}
+
+// NewFakeResponseWriter creates a new FakeResponseWriter that can be used to store the response a [http.Handler] made
+func NewFakeResponseWriter() *FakeResponseWriter {
```

Comment:
```suggestion
func newFakeResponseWriter() *FakeResponseWriter {
```

Reply:
Applied in db73a1f13bf8855a9c1a351a685f38629d5926bc thanks 😄